modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusManager.scala

```diff
@@ -38,134 +43,150 @@ object ConsensusManager {
+                .handleErrorWith(logger.error(_)(s"Error triggering consensus with time trigger"))
+            }.void
+        }
+
+      def checkForStateUpdate(key: Key)(resources: ConsensusResources[Artifact]): F[Unit] =
+        Spawn[F].start {
+          internalCheckForStateUpdate(key, resources)
+            .handleErrorWith(logger.error(_)(s"Error checking for consensus state update {key=${key.show}}"))
         }.void
-      }
-
-    private def internalCheckForStateUpdate(
-      key: Key,
-      resources: ConsensusResources[Artifact]
-    ): F[Unit] =
-      consensusStateUpdater.tryUpdateConsensus(key, resources).flatMap {
-        case Some(state) =>
-          state.status match {
-            case Finished(signedArtifact, majorityTrigger) =>
+
+      def checkForStateUpdateSync(key: Key)(resources: ConsensusResources[Artifact]): F[Unit] =
+        internalCheckForStateUpdate(key, resources)
+
+      private def internalFacilitateWith(
+        trigger: Option[ConsensusTrigger]
+      ): F[Unit] =
+        consensusStorage.getLastKeyAndArtifact.flatMap { maybeLastKeyAndArtifact =>
+          maybeLastKeyAndArtifact.traverse {
+            case (lastKey, Some(lastArtifact)) =>
+              val nextKey = lastKey.next
+
               consensusStorage
-                .tryUpdateLastKeyAndArtifactWithCleanup(state.lastKey, key, signedArtifact)
-                .ifM(
-                  afterConsensusFinish(majorityTrigger),
-                  logger.info("Skip triggering another consensus")
-                )
-            case CollectingProposals(_, None) =>
-              nodeStorage.tryModifyState(Observing, Ready) >>
+                .getResources(nextKey)
+                .flatMap { resources =>
+                  logger.debug(s"Trying to facilitate consensus {key=${nextKey.show}, trigger=${trigger.show}}") >>
+                    consensusStateUpdater.tryFacilitateConsensus(nextKey, lastKey, lastArtifact, trigger, resources).flatMap {
+                      case Some(_) =>
+                        internalCheckForStateUpdate(nextKey, resources)
+                      case None => Applicative[F].unit
+                    }
+                }
+            case _ => Applicative[F].unit
+          }.void
+        }
+
+      private def internalCheckForStateUpdate(
+        key: Key,
+        resources: ConsensusResources[Artifact]
+      ): F[Unit] =
+        consensusStateUpdater.tryUpdateConsensus(key, resources).flatMap {
+          case Some(state) =>
+            state.status match {
+              case Finished(signedArtifact, majorityTrigger) =>
+                consensusStorage
+                  .tryUpdateLastKeyAndArtifactWithCleanup(state.lastKey, key, signedArtifact)
+                  .ifM(
+                    afterConsensusFinish(majorityTrigger),
+                    logger.info("Skip triggering another consensus")
+                  ) >> nodeStorage.tryModifyStateB(Observing, Ready).void
```

Comment:
`nodeStorage.tryModifyStateB(Observing, Ready).void`
isn't it like `nodeStorage.tryModifyState(Observing, Ready)`?

Reply:
Also, I don't like the name of this function `tryModifyStateB`, but I couldn't come up with a better one.

Reply:
got it :+1: thanks

Reply:
maybe `tryModifyIfState(expected: State, to: State)`?

Reply:
I'm not convinced by this name, since the other function also modifies state if `actual == expected` and this name would suggest that only this function is doing such a check. 

Essentially it behaves exactly the same as `tryModifyState` but returns `Boolean` instead of raising an error.